File: 0146-lru-cache/0146-lru-cache.java
public class Node {
    public int key;
    public int val;
    public Node prev;
    public Node next;

    public Node(int key, int val) {
        this.key = key;
        this.val = val;
        this.prev = null;
        this.next = null;
    }
    
}

class LRUCache {
    private int capacity;
    private Map<Integer, Node> hashmap = new HashMap<>();
    private Node left;
    private Node right;
    
    public LRUCache(int capacity) {
        this.capacity = capacity;
        left = new Node(0, 0);
        right = new Node(0, 0);
        
        left.next = right;
        left.prev = null;
        
        right.next = null;
        right.prev = left;
    }
    
    private void remove(Node node) {
        // remove node from current pos in list
        Node prev = node.prev;
        Node next = node.next;
        
        prev.next = next;
        next.prev = prev;
    }
    
    private void append(Node node) {
        // add node to right of list
        Node temp = right.prev;
        right.prev = node;
        node.next = right;
        node.prev = temp;
        temp.next = node;
    }
    
    public int get(int key) {
        System.out.println("getting " + key);
        
        if (!hashmap.containsKey(key)) {
            return -1;
        }
        
        Node node = hashmap.get(key);
        
        // remove node from current pos in list
        remove(node);
        
        // add node to right of list
        append(node);
        
        return node.val;
    }
    
    public void put(int key, int value) {
        if (hashmap.containsKey(key)) {
            // key already exists, so remove old one
            remove(hashmap.get(key));
            hashmap.remove(key, hashmap.get(key));
            capacity++;
        }
        
        // add key to hashmap and add node to list
        Node node = new Node(key, value);
        hashmap.put(key, node);
        append(node);
        capacity--;
        
        System.out.println("putting " + key + ", capacity now " + capacity);
        // remove LRU if capacity exceeded
        if (capacity < 0) {
            Node lru = left.next;
            remove(lru);
            hashmap.remove(lru.key, lru);
            capacity++;
            System.out.println("removed " + lru.key);
        }
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */

File: 0146-lru-cache/0146-lru-cache.py
class Node:
    def __init__(self, key, val):
        self.key = key
        self.val = val
        self.prev = None
        self.next = None

class LRUCache:
    # approach: custom doubly-linked list, hashmap of key : node,
    # dummy left and right nodes

    def __init__(self, capacity: int):
        self.capacity = capacity
        self.size = 0
        self.hashmap = {}
        
        self.left = Node(0, 0)
        self.right = Node(0, 0)
        self.left.next = self.right
        self.right.prev = self.left

    def get(self, key: int) -> int:
        if key not in self.hashmap:
            return -1
        
        node = self.hashmap[key]
        self.delete(node)
        self.insert(node)
        
        return self.hashmap[key].val

    def insert(self, node: Node) -> None:
        temp = self.right.prev
        self.right.prev = node
        temp.next = node
        node.prev = temp
        node.next = self.right
        
    def delete(self, node: Node) -> None:
        next, prev = node.next, node.prev
        prev.next = next
        next.prev = prev
    
    def put(self, key: int, value: int) -> None:
        if key in self.hashmap:
            self.delete(self.hashmap[key])
            self.size -= 1
            
        node = Node(key, value)
        self.hashmap[key] = node
        self.insert(node)
        self.size += 1

        if self.size > self.capacity:
            lru = self.left.next
            lru.next.prev = self.left
            self.left.next = lru.next
            self.size -= 1
            del self.hashmap[lru.key]
        
# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)